packages/gatsby/src/redux/nodes.ts

```diff
@@ -221,14 +228,6 @@ const ensureIndexByTypedChain = (
       return
     }
 
-    // Special case `id` as that bucket never receives more than one element
-    if (chained === `id`) {
-      // Note: this is not a duplicate from `nodes` because this set only
-      //       contains nodes of this type. Page nodes are a subset of all nodes
-      byKeyValue.set(v, node)
-      return
-    }
```

Comment:
This bit of code was causing issues with the type system. Effectively what this is saying is that the type signature of `typedKeyValueIndexes` should be:

```ts
Map<string, Map<string | number | boolean, INode>> | Map<string, Map<string | number | boolean, Set<INode>>>
```

If that's not clear the difference is that the value of the nested Map can either be `INode | Set<INode>`. There is no logical way to type this as doing a union here means downstream code has to be aware that either type could exist (not knowing that it only applies if they key is 'id'). It's a heuristic the typesystem can't follow. So ultimately this optimization makes it harder to get the guarantees TS wants to give us.

I'm not a Perf expert, but it seems like a rather small optimization to have this flow anyhow. All I see is that it prevents an allocation of a Set, and prevents a spread (in the consuming file)

Reply:
cc @pvdz

Reply:
The effect is that this creates N more long-term `Set` instances, where N is the number of pages. So at scale this creates thousands of instances that serve no purpose. The spread is a one time (times N) cost, probably fine.

We would have to bench it to confirm the impact but I'm a little hesitant to go this route for the sake of pleasing TS.